### ISSUE TITLE
Download options are not working

### ISSUE THREAD

**Issue by seriousam** [YePpHa/YouTubeCenter#2334]
Forgive me if I do something wrong, or if someone else has already submitted for this issue.
Browser Mozilla firefox 51.0.1 64bit
Youtube center v538 (userscript)

If I try to download a video I just get a new(empty) tab, with a link similar to this: 
https://r7---sn-4vguioxu-n3be.googlevideo.com/videoplayback?itag=22&key=yt6&mime=video%2Fmp4&ei=erWPWLzUCpqO1gKXsZnYCg&source=youtube&upn=GJsXZfIt1pU&initcwndbps=757500&mn=sn-4vguioxu-n3be&mm=31&pl=18&id=o-AN1aO3dIHa2rhInar5nrLJl68A3QRLgtoUFTSNBlCsRJ&mv=m&sparams=dur%2Cei%2Cgcr%2Cid%2Cinitcwndbps%2Cip%2Cipbits%2Citag%2Clmt%2Cmime%2Cmm%2Cmn%2Cms%2Cmv%2Cpl%2Cratebypass%2Crequiressl%2Csource%2Cupn%2Cexpire&mt=1485813024&ms=au&requiressl=yes&ip=79.103.185.221&expire=1485834714&ratebypass=yes&dur=163.236&ipbits=0&gcr=gr&lmt=1474183657274784&title=Sound%20Of%20Legend%20-%20Blue%20(Da%20Ba%20Dee)%20(Audio)&cpn=JTcFjjTDa0K3Gg4-&signature=779BEB5E9434FB716ED14ECE558D558E522D56628.C9DB563C42DAD27C3EAD846DC4AE453BC3C0DBD5D5

**Comment by JMVS** [comment]
I've been investigating this issue and seems to be related to the URL encoding:
YouTubeCenter (not working): 
https://r5---sn-uxaxjxougv-x1x6.googlevideo.com/videoplayback?ipbits=0&lmt=1472154668096670&requiressl=yes&sparams=**dur%2Cei%2Cid%2Cinitcwndbps%2Cip%2Cipbits%2Citag%2Clmt%2Cmime%2Cmm%2Cmn%2Cms%2Cmv%2Cpl%2Cratebypass%2Crequiressl%2Csource%2Cupn%2Cexpire**&key=yt6&initcwndbps=1222500&mt=1486422024&pl=22&mv=m&ei=OwCZWNXvItfswgTwlquYAg&id=o-AM5RhwJc8yW_uIr-93E7h8pEmrGOyR2lt68W6tpb3oof&ms=au&mm=31&source=youtube&mn=sn-uxaxjxougv-x1x6&ip=186.23.76.63&ratebypass=yes&itag=22&expire=1486443675&upn=NoIzFATd2ug&dur=180.047&mime=video%2Fmp4&title=KURA%20%20NAMEK%20Official%20Music%20Video&cpn=4k-5VLYqNJIIvCBL&signature=FBFB06DD1D2EA6298F7F4671C09A8332F7261893C6.6CD1FD5955A0A477420CCB2827746449CC338743743

Another extension (working)
https://r5---sn-uxaxjxougv-x1x6.googlevideo.com/videoplayback?ipbits=0&lmt=1472154668096670&requiressl=yes&sparams=**dur,ei,id,initcwndbps,ip,ipbits,itag,lmt,mime,mm,mn,ms,mv,pl,ratebypass,requiressl,source,upn,expire**&key=yt6&initcwndbps=1222500&mt=1486422024&pl=22&mv=m&ei=OwCZWNXvItfswgTwlquYAg&id=o-AM5RhwJc8yW_uIr-93E7h8pEmrGOyR2lt68W6tpb3oof&ms=au&mm=31&source=youtube&mn=sn-uxaxjxougv-x1x6&ip=186.23.76.63&ratebypass=yes&itag=22&expire=1486443675&upn=NoIzFATd2ug&dur=180.047&mime=video/mp4&signature=347833CC94464772F2BCC024774A0A5595DF1DC6.6C3981627F2338A90C1764F7F8926AE2D1DD60B8&title=KURA%20-%20NAMEK%20(Official%20Music%20Video)

**Comment by Yonezpt** [comment]
This problem has already been detailed here: https://github.com/YePpHa/YouTubeCenter/issues/2098#issuecomment-157497362

**Comment by CollinChaffin** [comment]
I too see the old issue you linked too above (YEARS old) but now I just installed build 540 DEV and the entire download option is GONE?!?  I see no notes or mention (for some reason this app doesn't get any real changelog in the DEV cycle for YEARS at at time) so forgive me I did search but see no mention that a decision had been made to totally remove all download capabilities and frankly as a long time user was surpised when I updated my DEV build (after reading the limited info in Github on the changes which basically means having to read code diffs to see what has changed) to find the buttons are now just gone.

Can whomever made the code change care to comment and enlighten us users on this change, please?  Is the change permanent, etc. or just pending regex adjustments?

**Comment by Yonezpt** [comment]
@CollinChaffin You can see who changed the code and what was changed in the commit history: https://github.com/YePpHa/YouTubeCenter/commits/master

**Comment by seriousam** [comment]
@CollinChaffin I just updated to Build 540 DEV userscript. And I have download options and they are finally working...

**Comment by seriousam** [comment]
@CollinChaffin  well, I was wrong. I have the option to download, but it works only in some videos...in the rest videos I keep getting the same problem.

**Comment by seriousam** [comment]
@JMVS none of your links works for me :)

**Comment by Yonezpt** [comment]
@JMVS I have removed your links because they contained private data related to you, namely your IP address.

@seriousam As I explained in the link I posted, some videos cannot be downloaded until the YTC download feature is updated. Also his links only work for him and only for a limited time.

**Comment by CollinChaffin** [comment]
@seriousam Can you paste any link here that DOES work?  I'd feel better just seeing the button appear once because I have opened dozens of youtube videos and have yet to find one that shows me the YTC download button, it's just gone all together.  Lights-out, and all other functions, seem to be there though and working at least what I have had the time to test.

**Comment by YePpHa** [comment]
See https://github.com/YePpHa/YouTubeCenter/commit/d96b5aef5968a052bee89f007e4679a0a4bfe869

**Comment by CollinChaffin** [comment]
Nice!  Thanks YePpHa!

**Comment by Yonezpt** [comment]
@YePpHa the regex2 will always fail, YouTube hasn't used that pattern for quite a while, see https://github.com/YePpHa/YouTubeCenter/issues/2098#issuecomment-157497362

The download will always fail for streams that need signature decoding since the regex2 is not up to date.

**Comment by seriousam** [comment]
even after updating to build #541
Not working https://www.youtube.com/watch?v=jlLPQ5KzJkQ
Working https://www.youtube.com/watch?v=exUaDtl05U8&t=0s

**Comment by CollinChaffin** [comment]
Agreed, not working at all.  @YePpHa can you re-open this until the regex can be updated?

**Comment by CollinChaffin** [comment]
Build 542 looks much better!  I've tested a few vids and download working great again!  Nice work!  Also, awesome work on the new/updated CRX build for Chrome/win, also working perfectly!

**Comment by seriousam** [comment]
I have the exact same problem as before...
Well I don't have any knowledge of programming but  [this](https://greasyfork.org/en/scripts/1317-download-youtube-videos-as-mp4/code) might help you somehow, I have just tested it and it works for me.

**Comment by CollinChaffin** [comment]
@seriousam Can you give a public vid URL to show it not working on youtube?  Are you sure you are updated to build 542?  Every vid I have tested that failed before is now working properly for me with 542.

**Comment by Yonezpt** [comment]
@CollinChaffin are you able to download this video, for example: https://www.youtube.com/watch?v=o_1aF54DO60

**Comment by seriousam** [comment]
I have uploaded a video, where you can see the problem 
https://www.youtube.com/watch?v=ltJJPx4fzkk

**Comment by Yonezpt** [comment]
@seriousam That's because that second video has encoded signature, which YTC is failing to decode due to the problem I mentioned, you will have to wait for @YePpHa to correct it.

**Comment by YePpHa** [comment]
Okay, I have now fixed it. I have also tested it with the provided videos.

**Comment by Yonezpt** [comment]
@YePpHa I can't confirm it for myself, but are you sure you corrected the problem? Your regex2 pattern remains unaltered, which was the main problem as far as I could see https://github.com/YePpHa/YouTubeCenter/commit/98c4f2148f465b44724c97bb5937c5b5fe96906b#diff-7f8e8fd9264ba1b0f6bad74a21abad37R12315

**Comment by CollinChaffin** [comment]
@seriousam and @Yonezpt with new build 543 just released, I can confirm I am able to download both of the above videos perfectly.  Be sure you are updated to 543.

**Comment by CollinChaffin** [comment]
@Yonezpt see my screenshot, you're not looking at correct changes.  Here's all the changes Jeppe just made with 543 that resolves this bug:

https://1drv.ms/i/s!AhSqjTYUuHPKgagNYydR01Y6tGLp8A

**Comment by Yonezpt** [comment]
@CollinChaffin I saw all the changes, but the regex2 was not modified once, which I assumed to be the main problem, however, since you confirmed that the download is working for the videos with encoded signatures then my point becomes irrelevant.

**Comment by seriousam** [comment]
I can approve that it works now (at least for a few videos I just tested).
Thank you all for interest ,time and effort.

**Comment by JMVS** [comment]
I can confirm it is working now. Thank you.

**Comment by YePpHa** [comment]
@Yonezpt It's actually only `regex` that's used. `regex2` is not being used at the moment (`regex` should technically be before `regex2`).

The problems that I fixed with the latest update were:
* The `base.js` file was changed from `//youtube.com/yts/jsbin/player-en_US-vflwAvMHx/base.js` to `/yts/jsbin/player-en_US-vflwAvMHx/base.js`. The old method of loading the base.js file expected that only the protocol was not defined. This caused it to try loading `https:/yts/jsbin/player-en_US-vflwAvMHx/base.js` instead of correctly adding the domain and protocol.
* The regex called `regex` couldn't handle new lines inside the function definitions in `base.js`.
